Issue 1232 - variadic is_nan function

andrjohns opened the pull request:
## Summary
As initially described in #1232 - Several functions call ```is_nan``` multiple times using the ```||``` operator, such as in ```prim/scal/fun/grad_reg_inc_gamma```:
```
  if (is_nan(a) || is_nan(z) || is_nan(g) || is_nan(dig))
    return std::numeric_limits<TP>::quiet_NaN();
```
This PR adds a variadic extension to the ```is_nan``` function so that multiple inputs can be passed in a single call
```
  if (is_nan(a, z, g, dig))
    return std::numeric_limits<TP>::quiet_NaN();
```

## Tests
Added tests for calling ```is_nan``` with multiple inputs

## Side Effects
N/A

## Checklist

- [X] Math issue #1232 

- [X] Copyright holder: Andrew Johnson

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

syclik commented on stan/math/prim/scal/fun/is_nan.hpp:
```diff
@@ -14,8 +15,24 @@ namespace math {
+}
 
+/**
+ * Returns true if the input is NaN and false otherwise.
```
Can you make this doc more precise? It should indicate that it returns `true` if any of the input is `NaN` (as opposed to all).

wds15 commented:
I think what @bob-carpenter said is important: The function should become `is_nan_any` or something similar.

andrjohns commented:
@syclik I've made the changes from the issue discussion and the tests have passed, this is ready for re-review

syclik commented on stan/math/prim/scal/fun/is_any_nan.hpp:
```diff
@@ -0,0 +1,41 @@
+ * @return <code>true</code> if the value is NaN.
+ */
+template <typename T>
+inline bool is_any_nan(T x) {
```
We should have caught this on the original PR that introduced `is_nan`. The arguments are passed by value here. They should be passed by reference for efficiency reasons. So... in the function signature: `is_any_nan(T& x)` instead of `is_any_nan(T x)`.

syclik commented on stan/math/prim/scal/fun/is_any_nan.hpp:
```diff
@@ -0,0 +1,41 @@
+ * @return <code>true</code> if any value is NaN
+ */
+template <typename T, typename... Ts>
+inline bool is_any_nan(T x, Ts... xs) {
```
Same here. This should be by reference.

syclik commented on stan/math/prim/scal/fun/is_nan.hpp:
```diff
@@ -14,7 +14,10 @@ namespace math {
  */
-inline bool is_nan(double x) { return std::isnan(x); }
+template <typename T>
+inline bool is_nan(T x) {
```
Hmm... this looks odd. In `develop`, we have 3 definitions of `is_nan`: prim/scal, rev/scal, fwd/scal. They are implemented using function overloading.

I think you either want to go the same route or you want to make this function the template function (as you have it) and then change the implementations in rev and fwd to be template specializations.

syclik replied:
Does that make sense? If it's not clear what I'm talking about, we can discuss.

syclik commented on stan/math/prim/scal/fun/log_falling_factorial.hpp:
```diff
@@ -54,7 +54,7 @@ namespace math {
 inline typename return_type<T1, T2>::type log_falling_factorial(const T1 x,
                                                                 const T2 n) {
-  if (is_nan(x) || is_nan(n))
+  if (is_any_nan(x, n))
```
[comment]
The changes look awesome! It does look like there's an impact on how we will write functions here on out.

andrjohns commented:
@serban-nicusor-toptal would you be able to restart the "Always-run tests part 1" for this pull? Looks like the failure was just randomness, as discussed in #1238

serban-nicusor-toptal commented:
@andrjohns Good morning!
I've restarted it, it's wating for a gpu machine. Keep me updated!

andrjohns commented:
Thanks for that!

andrjohns commented:
@serban-nicusor-toptal looks like the build is running twice for some reason (there's a run 8 and a run 9), not sure what happened there

rok-cesnovar commented:
I got the email and restarted it. Should have posted here sorry.

andrjohns commented:
No worries, I'm not about to complain for having too much help!

andrjohns commented:
@syclik This is ready for another look-over. I've updated ```is_any_nan``` to pass by reference and changed the rev/fwd versions of ```is_nan``` to template specialisations.

I also added in fwd & rev tests for ```is_any_nan```, just to be safe

bob-carpenter commented on stan/math/fwd/scal/fun/is_nan.hpp:
```diff
@@ -2,21 +2,23 @@
  */
 template <typename T>
-inline int is_nan(const fvar<T>& x) {
+inline typename std::enable_if_t<is_fvar<T>::value, bool> is_nan(const T& x) {
```
What was the motivation for switching to `enable_if_t` from the simple pattern matching?  I think this signature should just be:

```
template <typename T>
inline bool is_nan(const fvar<T>& x);
```

andrjohns replied:
This was one was based on @syclik's review:
> Hmm... this looks odd. In develop, we have 3 definitions of is_nan: prim/scal, rev/scal, fwd/scal. They are implemented using function overloading.

> I think you either want to go the same route or you want to make this function the template function (as you have it) and then change the implementations in rev and fwd to be template specializations.

bob-carpenter commented on stan/math/prim/scal/fun/is_any_nan.hpp:
```diff
@@ -0,0 +1,41 @@
+ */
+template <typename T, typename... Ts>
+inline bool is_any_nan(const T& x, const Ts&... xs) {
+  return (is_any_nan(x) || is_any_nan(std::forward<const Ts>(xs)...));
```
These are both OK to be constant references so they'll be generic.  You don't need those outer parens---that's an R thing, not a C++ thing and they're just a distraction here.

bob-carpenter commented on stan/math/prim/scal/fun/is_nan.hpp:
```diff
@@ -14,7 +15,11 @@ namespace math {
  */
-inline bool is_nan(double x) { return std::isnan(x); }
+template <typename T>
+inline typename std::enable_if_t<std::is_arithmetic<T>::value, bool> is_nan(
```
Same comment here as in the issue for changing all of these.  At most, this should be

```
template <typename T>
inline typename std::enable_if_t<std::is_arithmetic<T>::value, bool> is_nan(T x) {
  return std::isnan(x);
}
```

so primitives get passed by value.

andrjohns commented:
@syclik This one is just waiting on your approval before it's ready to merge, thanks!

bob-carpenter commented:
I've never been clear on where we should use template specializations and where we should just use overloads.  Name resolution works by finding the best matching (template) function overload, then finding the best matching specialization.  So I see how it works differently, just not when it would make a difference in a real situation.

> On May 26, 2019, at 9:38 PM, Daniel Lee <notifications@github.com> wrote:
> 
> @syclik commented on this pull request.
> 
> @andrjohns: most of it is good. I think there was confusion about how is_nan() should be implemented. I have the same question that @bob-carpenter had. Why do we need the enable_if_t at all?
> 
> In prim, it could be:
> 
> template <typename T>
> inline bool is_nan(T x) {
>    return std::isnan(x);
> }
> 
> in fwd:
> 
> template <>
> template <typename T>
> inline bool is_nan(const fvar<T>& x) {
>    return is_nan(x.val());
> }
> 
> and rev:
> 
> template <>
> inline bool is_nan(const var& x) {
>   return is_nan(x.val());
> }
> 
> Is there something wrong with the approach above? In the Math library, where we can, we just use the template specializations directly rather than trying to do clever trait programming. Simpler is better for reading this code.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

syclik commented:
@bob-carpenter: I'm with you. I sort of view them as exchangeable for this sort of application (although someone might be able to tell me why one would be preferred). What I was trying to get at with my comments was that we don't need the `enable_if` type of construction for this use case. It's just easier to read and follow when it's less complex.

andrjohns commented:
@syclik @bob-carpenter 
Should I just leave ```is_nan``` for autodiff types unchanged from develop? Seems like there's no preference either way, so I don't think there's much need to change it

bob-carpenter commented:
Fine by me, but it's Daniel's call.

drezap commented:
@syclik it's your call ^^^^^ (just trying to keep things moving. this is a good change).

rok-cesnovar commented:
@andrjohns when you have the time, please merge the latest develop in. There are conflicts due to the meta flattening refactor.

andrjohns commented:
@syclik Did you have a preference on the specification of ```is_nan``` for autodiff types (i.e. leave unchanged from develop or ```enable_if``` templating)?

syclik commented:
My preference is for simpler rather than more complex. Here, I'd prefer develop rather than using `enable_if`.

andrjohns commented:
Great! Changes have been reverted

andrjohns commented:
@serban-nicusor-toptal If the windows threading errors aren't an issue anymore can you restart the testing for this one? Thanks
